src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,713 @@
+  new_name_(config.get_new_name(key_)) {
+  // Worker constructor - try to get the binary written to local file
+  if (is_global_) return;
+  int rc = 0;
+  config.ensure_module_dir_exists();
+  auto mylock = std::make_shared< std::lock_guard<NativeModule> >(*this);
+  for (;;) {
+    struct stat lib_stat;
+    if (file_stat(lib_name_, &lib_stat) && (lib_stat.st_size == binary_size)) {
+      build_state_ = kPass;
+      break;
+    }
+    // Race to write the new file
+    int fd = open(new_name_.c_str(), O_WRONLY|O_CREAT|O_EXCL, 0666);
+    if (fd >= 0) {
+      // Now we're about to write the new file
+      rc = write(fd, binary, binary_size);
+      assert(rc == binary_size);
+      ::close(fd);
+      ::chmod(new_name_.c_str(), 0644);
+      struct stat st;
+      if (file_stat(lib_name_, &st)) {
+        long old_size = st.st_size;
+        if (old_size == binary_size) {
+          ::unlink(new_name_.c_str());
+          break;
+        } else if (old_size == 0) {
+          ::unlink(lib_name_.c_str());
+        } else {
+          auto old = lib_name_ + ".old";
+          ::rename(lib_name_.c_str(), old.c_str());
+        }
+      }
+      // Don't let anyone see the file until it is completely written
+      rc = ::rename(new_name_.c_str(), lib_name_.c_str());
+      build_state_ = ((rc == 0) ? kPass : kFail);
+      break;
+    } else {
+      // Someone else is writing to new
+      while (file_exists_and_is_recent(new_name_)) {
+        usleep_without_lock(kFilePollMicrosecs);
+      }
+    }
+  }
+  mylock.reset();
+  if (build_state_ == kPass) {
+    try_load();
+  }
+}
+
+NativeModule::~NativeModule() {
+  if (!is_global_ && dlopen_handle_) {
+    dlclose(dlopen_handle_);
+  }
+}
+
+// We may have many threads across many processes trying to access the shared
+// files in ~/hail_modules.  We protect the files for each module by
+// mutual exclusion based on the existence of a link hm_${key}.lock
+//
+// To achieve disaster recovery after a killed process or crash, a .lock
+// link will be ignored if it is more than 20 seconds old.  The critical
+// sections protected by the lock should be much shorter than this.
+//
+// Permission to build a new file is controlled by the existence of hm_${key}.new,
+// and has a longer timeout (120 sec) to allow for module compile time.
+//
+// The intent is that when hail is not running, it should be ok to do
+// "rm ~/hail_modules/*" to clear the cache and start again.
+
+void NativeModule::lock() {
+  auto target = (lock_name_ + std::string("X"));
+  for (;;) {
+    // Creating a link is atomic
+    errno = 0;
+    int rc = ::link(target.c_str(), lock_name_.c_str());
+    int e = errno;
+    if (rc == 0) {
+      break;
+    } else if (e == EEXIST) { // Link already existed
+      struct stat st;
+      rc = ::stat(lock_name_.c_str(), &st);
+      if ((rc == 0) && (st.st_ctime+20 < ::time(nullptr))) {
+        fprintf(stderr, "WARNING: force break old lock %s\n", lock_name_.c_str());
```

Comment:
in our scala code the log format looks like:
```
2018-07-11 20:15:15 Hail: INFO:
```
It'd be nice if we could keep this consistent. Would it be prohibitively expensive to format date & time for an error log?

Reply:
I could do that.  I'm not sure C++ stderr is going to end up where we want it though.  Improving
the mixed-language infrastructure around debugging, logging, and IO (allowing C++ to access
cloud object storage and HDFS) is going to need work over the next month or two.